### PR TITLE
Fix typing in shapes helper files

### PR DESCRIPTION
**Pull request by dstansby**
# Description
Fix typing in `napari.layers.shapes.{_shape_list, _shapes_key_bindings, _shapes_utils}`. Also fixes some typing in other shapes files, but there are lots of errors in `shapes.py` and `_shapes_mouse_bindings.py` so I think best to leave those to a follow up PR.

## Type of change

- [x] Maintenance (changes required to run napari, tests, & CI smoothly)

### CODE REVIEW

File: napari/layers/shapes/_shape_list.py [hunk 1]
```diff
@@ -69,11 +69,11 @@ class ShapeList:
 
     def __init__(self, data=(), ndisplay=2) -> None:
         self._ndisplay = ndisplay
-        self.shapes = []
-        self._displayed = []
-        self._slice_key = []
```
Comment (Carreau) [inline napari/layers/shapes/_shape_list.py:89]: This is triggering a failure in test. 

This is dues to numpy's `==` that have deprecated comparison of differently sized arrays. I belive you now need to use `np.array_equal`

Comment (Carreau) [reply napari/layers/shapes/_shape_list.py:89]:
```diff
diff --git a/napari/layers/shapes/_shape_list.py b/napari/layers/shapes/_shape_list.py
index 3677b5a15..11861d4f7 100644
--- a/napari/layers/shapes/_shape_list.py
+++ b/napari/layers/shapes/_shape_list.py
@@ -187,7 +187,7 @@ class ShapeList:
     @slice_key.setter
     def slice_key(self, slice_key):
         slice_key = list(slice_key)
-        if not np.all(self._slice_key == slice_key):
+        if not np.array_equal(self._slice_key, slice_key):
             self._slice_key = slice_key
             self._update_displayed()
```
Seem to fix most of the failure, but seem to actually create some genuine issues down the road, which might be hidden bugs, or broadcast issues.

Comment (Carreau) [reply napari/layers/shapes/_shape_list.py:89]: Ok, there is a difference in broadcasting: 

```
In [5]: np.array_equal(np.array([[1,2],[1,2]]),  np.array([1,2]))
Out[5]: False

In [6]: np.all(np.array([[1,2],[1,2]]) == np.array([1,2]))
Out[6]: True
```

File: napari/layers/shapes/shapes.py [hunk 2]
```diff
@@ -1583,7 +1595,7 @@ def _view_text_coords(self) -> Tuple[np.ndarray, str, str]:
         ]
 
         return self.text.compute_text_coords(
-            sliced_in_view_coords, ndisplay, order
+            np.asarray(sliced_in_view_coords), ndisplay, order
         )
```
Comment (Carreau) [inline napari/layers/shapes/shapes.py]: `compute_text_coords`,  `calls get_text_anchors`, which dependings on it's parameters calls `_calculate_anchor_center`, itself using `_calculate_bbox_centers`. 

_calculate_bbox_centers behavior depends on wether it is given a  list or an ndarray, and create a failure. 

as `_calculate_bbox_centers` seem to be only used for this specific purpose I'm tempted to assume this is a bug in `_calculate_bbox_centers` which behavior should be fixed, likely converted from list to ndarray or vice-versa higher up in the stack.

My take is to mark this specific one right here with a temporary `# type: ignore`. and have a gh issue for it

**Comment by Carreau** [comment]
I've pushed a couple of changes that should make CI pass. Marking as ready to review.

File: napari/layers/shapes/shapes.py [hunk 3]
```diff
@@ -503,13 +514,22 @@ def __init__(
...
         self._last_cursor_position = None
 
+        self._face_color_property: str
+        self._edge_color_property: str
+        self._edge_color_cycle: npt.NDArray
+        self._edge_color_cycle_values: npt.NDArray
+        self._face_color_cycle: npt.NDArray
+        self._face_color_cycle_values: npt.NDArray
+        self._edge_color_mode: str
+        self._face_color_mode: str
```
Comment (Carreau) [inline napari/layers/shapes/shapes.py]: Can/should all those be typed as the class-level ?

**Comment by melonora** [comment]
Sorry that this one takes a while to get back too. I will try to have a look later today.

**Review by Carreau — APPROVED** [review]

**Pull request merged**